stdlib/Random/src/Random.jl

```diff
@@ -434,4 +434,9 @@ true
 """
 seed!(rng::AbstractRNG, ::Nothing) = seed!(rng)
 
+# Randomize quicksort pivot selection. This code is here because of bootstrapping:
+# we need to sort things before we load this standard library.
+# TODO move this into Sort.jl
+Base.Sort.select_pivot(lo::Integer, hi::Integer) = rand(lo:hi)
```

Comment:
Would it be better to use something other than the global RNG? Perhaps introduce a dedicated stream for sorting operations? I certainly wouldn't anticipate that `sort` would affect the global RNG and it would take me a while to trace the source of the consumption to here. Technically, a user shouldn't care... but that isn't to say they won't.

Reply:
I don't think we should try to make the promise that algorithms in `Base`/`Stdlib` won't generate random numbers using the global rng. A ton of stuff needs random numbers (eg `median`, some matrix factorization algorithms (for pivoting), possibly thread scheduling). I think we shouldn't set the expectation that users should care about the global RNG state.

Reply:
While it would be nice to not alter the global RNG seed, to maintain our "everything is task scheduling independent, even random numbers" guarantee, which, I believe, is much more important than "sorting doesn't change global RNG state", I think we would need to have a different dedicated stream for each task, which is far too expensive; or we would need to branch a new stream from the global RNG without altering the global RNG's state, which could introduce very hard to anticipate correlations between the randomness of sorting and of other generations akin to but much subtler (and less likely to come up) than https://github.com/JuliaLang/julia/issues/6573

In general, I think everything should use the global RNG, and if someone wants a stream nothing else touches, then they are responsible for creating their own dedicated stream.

There is a warning at the top of https://docs.julialang.org/en/v1/stdlib/Random/ that is somewhat relevant, though it doesn't directly apply to this.

Reply:
Beat me to it :) I totally agree.

Reply:
This seems to have caused a bug in Zygote: https://github.com/FluxML/Zygote.jl/issues/1351. Possibly this should just be a pure Zygote fix, but I thought it would be good to be aware that this has the potential to cause some bugs in cases where this function was assumed not to affect the RNG state.

Reply:
A hidden RNG doesn't seem ideal to me because it feels like a good invariant is that a) the global RNG's initial state, and b) any user-provided RNGs make a program'd output deterministic. Hidden RNGs break that; now the user needs to know about them too.

Of course, the output of `sort!` is ultimately deterministic, which helps. But the computation path still isn't: stick an accumulator into the `by` function and now there's observable behavior dependent on the hidden RNG.

That being said, I really have no idea what the best solution to #48230 is, assuming people agree it's a problem. Perhaps all the call points within the internals (REPL etc.) as well as any "ex-internals" (Zygote etc.) would all need to be fixed to use a sorting algorithm that guarantees no random calls, but that's a sizable effort. 

I'll leave it up to people who really know what they're talking about to decide, but just trying to point out the issues as I see it here:)

Reply:
Forking a new RNG off of the default RNG would be a possibility and just feeding it through all the internal `sort!` calls. We don't really need good randomness for sorting, just something that won't accidentally do something with a pattern that matches the data.

Reply:
Modifying the internal calls does seem reasonable. And some effort would then need to made to find cases like the Zygote one in the ecosystem.

As for how to modify the calls, forking the RNG and using that as you suggest seems reasonable (I don't believe the `sort!` currently accepts user provided RNGs, so that would also need to be fixed, which should probably happen anyway), although just using e.g. `MergeSort` and adding a documented guarantee that the algorithm is purely deterministic seems simpler to me? That is, unless the speed of `QuickerSort` is needed for the internal calls.

Reply:
Using `hash(hi) % (hi-lo+1) + lo` instead of `rand(lo:hi)` would be the easiest solution and should be robust against pretty much all non-malicious input. In 1.8 we also had pathological (and more easily constructed) quadratic cases so this would not be a regression.

This would also remove sorting's interaction with https://github.com/JuliaLang/julia/pull/48069 entirely.

Reply:
that would be reasonable.

Reply:
That seems reasonable to me? I still think something should be done about the "does sort! semantically guarantee determinism" question because if the answer is no then it's internal usage in various places is incorrect.

But perhaps it would be good in any case to have a version cycle to mull over that? And do the quick fix you suggest for now, since it's not a regression as you say.

Reply:
it depends what you mean by determinism.

Reply:
What I mean in my usage is: can `sort!` affect the global RNG state? (As it should, if it makes calls to rand and does not use a hidden RNG which is bad for other reasons.)

Likely, the answer should be "yes, it can", even if it doesn't in 1.8 and now won't in 1.9 with the reversion to hash. In which case, the semantics of the repl+zygote are messed up. So their calls to `sort!` need the answer to the above question to be a guaranteed no (either by using a "deterministic" alg, or by using a special user-provided RNG to a "random" alg).

Reply:
> it's internal usage in various places is incorrect.

Do you have an example?

> In 1.8 we also had pathological (and more easily constructed) quadratic cases

For example, the "sawtooth" ordering:

```julia
julia> n = 10^6;

julia> x = vcat(3n:4n,2n:3n,n:2n,0:n);

julia> @time sort!(shuffle!(x));
  0.477350 seconds

julia> x = vcat(3n:4n,2n:3n,n:2n,0:n);

julia> @time sort!(x);
134.471621 seconds

julia> VERSION
v"1.8.5"
```

Reply:
I don't have a clean MWE, but there are two clunkier examples in https://github.com/JuliaLang/julia/issues/48230 (Zygote is one breakage, and then the repl which is a theoretical one since we probably get a dispatch to insertionsort in practice)

Reply:
But you could imagine it could have a lot of consequences in various non-barebones environments where `sort!` is used incautiously, e.g. maybe in Pluto notebook, for whatever reason a `sort!` is run when processing a cell output (this is purely hypothetical). Ideally we'd want the same code with the same seed to run the same in a these environments, which would be in danger if `rand` is introduced to `sort!` without carefully examining such cases

Reply:
The default rng should not be used for reproducible random streams.

Reply:
It's very common practice to begin a piece of code with `Random.seed!(...)` to make it reproducible. Faulty a strategy as it may be, it would be confusing if calls to `sort!` made by non user -facing code broke this. That's why I think those calls should eventually be changed to something which can be relied not to affect the global RNG.


In any case, I don't want to debate this point too much here since I'm not pushing for this to affect 1.9.